### PR TITLE
Add `MonitorHandle::current_video_mode()`

### PR DESCRIPTION
**Pull request by daxpedda**
After some discussion we decided its more useful to encode current monitor information into `VideoModeHandle` then move more information to `MonitorHandle`. To this end this PR implements `MonitorHandle::current_video_mode()`.

Additionally `MonitorHandle::size()` and `refresh_rate_millihertz()` were removed, because it logically fits into video modes and not the monitor itself.

To improve the output, this PR also changes `VideoModeHandle::refresh_rate_millihertz()` to return a `Option<NonZeroU32>` and `size()` and `bit_depth()` an `Option`.

Follow up to #3801.

### CODE REVIEW

File: src/platform_impl/apple/appkit/monitor.rs
Comment (daxpedda) [inline src/platform_impl/apple/appkit/monitor.rs:249]: Please help @madsmtm :cold_sweat:.
Not sure why this is unsafe or if there is a better way to do this.
I guess I could also just assume that the number of color components is 4?

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:1016]: On Android this seems hard. We could ask for the current format but various APIs can always change it.

I've seen though that after changing the Surface/window format, APIs like EGL suddenly return just the configs matching that exact format.

Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1016]: > We could ask for the current format but various APIs can always change it.

I think it should be clear to users that the data only reflects the state at the time it was queried.

In the future it would be nice if we could have some events that reflect a change in available monitors or their settings. Its definitely on my checklist for Web.

> On Android this seems hard.

I'm not too worried about it in this PR because `VideoModeHandle` on Android already always returns 32, so this definitely doesn't change anything about the current state of things.

**Comment by kchibisov** [comment]
The `bit_depth` is more of a property of the active video format, so if I'd change the way handle works I'd probably just keep a handle and an option to get the active `VideoMode` for stuff like that.

The only information which is not video mode related is basically monitor name/position, etc, depth is not it.

**Comment by daxpedda** [comment]
> The `bit_depth` is more of a property of the active video format, so if I'd change the way handle works I'd probably just keep a handle and an option to get the active `VideoMode` for stuff like that.
> 
> The only information which is not video mode related is basically monitor name/position, etc, depth is not it.

That's a good point. That would solve the issue for Web as well.
Will look into it and adapt the PR.

**Comment by daxpedda** [comment]
> The `bit_depth` is more of a property of the active video format, so if I'd change the way handle works I'd probably just keep a handle and an option to get the active `VideoMode` for stuff like that.
> 
> The only information which is not video mode related is basically monitor name/position, etc, depth is not it.

@kchibisov I started working on it right now and realized that this argument invalidates `MonitorHandle` currently having size and refresh rate. Is there some story behind it or should we remove it in favor of `current_video_mode()` exposing the current size and refresh rate?

**Comment by kchibisov** [comment]
>Is there some story behind it or should we remove it in favor of current_video_mode() exposing the current size and refresh rate?

I think it was always like that, which doesn't make much sense. Probably it was like that for convenience or something. Just route everything into `current_video_mode`.

File: src/platform_impl/apple/appkit/monitor.rs
Comment (daxpedda) [reply src/platform_impl/apple/appkit/monitor.rs]: Resolved by the PR completely changing.

**Comment by daxpedda** [comment]
> > Is there some story behind it or should we remove it in favor of current_video_mode() exposing the current size and refresh rate?
> 
> I think it was always like that, which doesn't make much sense. Probably it was like that for convenience or something. Just route everything into `current_video_mode`.

Done.

File: src/changelog/unreleased.md
Comment (daxpedda) [reply src/changelog/unreleased.md:74]: I like it!

Comment (daxpedda) [reply src/changelog/unreleased.md]: Done.

**Review by kchibisov — APPROVED** [review]

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:1037]: Would you mind adding a TODO that we should get this via either https://developer.android.com/reference/android/view/Display#getRefreshRate() (requires JNI, but works all the way back to API level 1) or via https://developer.android.com/ndk/reference/group/choreographer#achoreographer_registerrefreshratecallback (only available since API level 30, related to https://developer.android.com/media/optimize/performance/frame-rate).

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:1010]: They seem to have [deprecated the ability to get the pixel format](https://developer.android.com/reference/android/view/Display#getPixelFormat()) long ago, unfortunately.

There's no point querying the format of the `Surface` since:
1. There won't be a `Surface` until the `Resumed` event;
2. The format typically starts out with `RGB565` for some arcane reason;
3. The user can freely change this pixel format and the compositor will Deal With It™.

Maybe this needs to be `Option` as well?

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:1016]: Leaving this here just in case: https://developer.android.com/reference/android/view/Display.Mode

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [reply src/platform_impl/android/mod.rs:1037]: Also curious what we should do in the case of VRR.

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:1046]: Perhaps we could pull this from the `configuration` instead, but it's all DPI-dependent:

https://docs.rs/ndk/latest/ndk/configuration/struct.Configuration.html#method.screen_width_dp

**Review by MarijnS95 — APPROVED** [review]
Approving the Android side given that all the weirdness in there (due to a lack of API in the NDK - we could call into Java for this though) was already there and is only moving in this patch but not fundamentally changing.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1010]: > 1. There won't be a `Surface` until the `Resumed` event;

Why would that make it pointless? Do you mean because when you have a surface you use your graphics API to determine your pixel format anyway?

In that case the API would be useless on all platforms wouldn't it?
Personally I don't know who would use that, so I would actually be in favor of removing it wholesale if we come to a consensus.

> Maybe this needs to be `Option` as well?

I think we should change it to `None` if its not supported.

My assumption was always that the main use-case here is to check for HDR support, which Android would support with [`android.view.Display.getSupportedHdrTypes()`](https://developer.android.com/reference/android/view/Display.Mode#getSupportedHdrTypes()). But that API should probably look different and again, not useful to handle in Winit. E.g., we discussed querying pixel format support in Softbuffer already https://github.com/rust-windowing/softbuffer/issues/98 and Wgpu obviously already supports that.

---

If you don't mind I would leave it like this here and make a follow up PR discussing any further changes.
WDYT in any case about all this?

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1054]: Let me know if you want me to add any comment for that.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1017]: Added it as a TODO comment.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1045]: Done!

> Also curious what we should do in the case of VRR.

Ha!
We would need to change the API for that, lol.

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [reply src/platform_impl/android/mod.rs:1010]: > > 1. There won't be a `Surface` until the `Resumed` event;
> 
> Why would that make it pointless? Do you mean because when you have a surface you use your graphics API to determine your pixel format anyway?

Pointless _on Android_, since I think you can query the video mode (and most apps probably will?) before a "surface" appears?

The pixel format of a surface can and will almost always be disjoint from the bit depth of the video mode on a physical panel/connection because of compositors (though it's probably beneficial to keep them in sync depending on what the scanout hardware is capable of).

> In that case the API would be useless on all platforms wouldn't it? Personally I don't know who would use that, so I would actually be in favor of removing it wholesale if we come to a consensus.

I think all platforms with a list of monitors and their current mode can perfectly benefit from this API. A pixel format may be more helpful over a plain bit depth though, but that's bikeshedding at this point.

> > Maybe this needs to be `Option` as well?
> 
> I think we should change it to `None` if its not supported.

Exactly: if this function can return `Option`, its value will always be `None` on Android.

> My assumption was always that the main use-case here is to check for HDR support, which Android would support with [`android.view.Display.getSupportedHdrTypes()`](https://developer.android.com/reference/android/view/Display.Mode#getSupportedHdrTypes()). 

You can't do that with bit-depth alone, e.g. RGBA8888 and R10G10B10A2 both use 32 bits. Both could be used for a HDR _color space_ but the former likely has visible banding.

> But that API should probably look different and again, not useful to handle in Winit. 

Yeah, HDR is much, _much_ more complicated than that. Better leave it to a users' Vulkan/Dxgi/(E)GL implementation.

Though if a platform provides additional HDR-related details for the current/negotiated _mode_ for a monitor via its native API, it _might_ be nice to expose that on the mode handle.
(things like HDR metadata, VRR, color space from the mode; you have that in Windows and DRM at least)

> E.g., we discussed querying pixel format support in Softbuffer already https://github.com/rust-windowing/softbuffer/issues/98 and Wgpu obviously already supports that.

Seems like that should be disconnected from winit, yes.

> If you don't mind I would leave it like this here and make a follow up PR discussing any further changes. WDYT in any case about all this?

:+1:

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [reply src/platform_impl/android/mod.rs:1054]: I think I/we should just play with that idea, but maybe for now let the size of the mode always return `None` or something, to disconnect it from a possible surface?

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1010]: > The pixel format of a surface can and will almost always be disjoint from the bit depth of the video mode on a physical panel/connection because of compositors (though it's probably beneficial to keep them in sync depending on what the scanout hardware is capable of).

...

> > In that case the API would be useless on all platforms wouldn't it? Personally I don't know who would use that, so I would actually be in favor of removing it wholesale if we come to a consensus.
> 
> I think all platforms with a list of monitors and their current mode can perfectly benefit from this API. A pixel format may be more helpful over a plain bit depth though, but that's bikeshedding at this point.

Well that would still only make sense if the bit depth has anything to do with the pixel format when you start drawing. Your earlier statement seems to say that this isn't a guarantee?

> > > Maybe this needs to be `Option` as well?
> > 
> > 
> > I think we should change it to `None` if its not supported.
> 
> Exactly: if this function can return `Option`, its value will always be `None` on Android.

Yeah, that's fine. Other platforms don't support querying it as well.

---

It seems this discussion largely boils down to bit depth vs pixel format. In Winit, the bit depth is only useful when you *choose* a video mode handle, which you can via `Window::set_fullscreen()` (not on Android though). But this all implies that the bit depth has some connection to the pixel format, obviously the bit depth doesn't contain all the information of a pixel format, but it does make a difference. E.g. on a 6-bit video mode you can't use 8-bit color pixel format.

In reality of course, the choice is usually between 8-bit (99.9%) and 10-bit per color. No clue how HDR plays a role here, but as we said this would require a different API.

Seems all pointless on Android though if you can't use that information in `Window::set_fullscreen()`, so this basically boils down to use just exposing device data.

---

Conclusion: make it an `Option`.
(will do in a follow-up)

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1054]: AFAIK `VideoModeHandle::size()` shouldn't be used to create surfaces anyway, they should just tell you the the resolution in physical pixels of that video mode.

I will add that to the documentation.
Is that enough or is the information exposed here just wrong?

Comment (MarijnS95) [reply src/platform_impl/android/mod.rs:1054]: Exactly that as well. The surface will barely ever be the fill resolution of the screen when statusbars and navbars are removed from the screen.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs]: We solved this by returning `None` here now.
See https://github.com/rust-windowing/winit/pull/3802#discussion_r1690481503 for more context.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs]: I've done it in this PR after all.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1047]: I've changed `VideoModeHandle::size()` to return an `Option` now.

**Comment by daxpedda** [comment]
Re-requested reviews because I changed `VideoModeHandle::size()` and `bit_depth()` to return an `Option` now.
Some backends obviously make use of that now instead of returning default values.

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [reply src/platform_impl/android/mod.rs]: > It seems this discussion largely boils down to bit depth vs pixel format.

Can't help when discussing how to implement this function, we look at where to pull the necessary information from, and eventually what the API even means and how it is useful. Let me have a look at `Window::set_fullscreen()` and how it utilizes the `bit_depth()` specifically.

> In reality of course, the choice is usually between 8-bit (99.9%) and 10-bit per color.

Again, it's hard to tell those apart if you have `R10G10B10A2` versus `R8G8B8A8` or `R8G8B8X8` which are both 32-bits. We'd need to return the _per-channel_ bit depth in that case?

Comment (daxpedda) [reply src/platform_impl/android/mod.rs]: > > In reality of course, the choice is usually between 8-bit (99.9%) and 10-bit per color.
> 
> Again, it's hard to tell those apart if you have `R10G10B10A2` versus `R8G8B8A8` or `R8G8B8X8` which are both 32-bits. We'd need to return the _per-channel_ bit depth in that case?

It depends if those are part of video modes. If they are pixel formats then I don't think the are relevant to Winit.

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:770]: Maybe we should rename this helper to `surface_size()` or inline it entirely?

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:1023]: As discussed we're contemplating whether to make it `Option` if we have a chance of falling back to JNI to implement this function - and as a consequence never return `None` on any platform.

**Review by MarijnS95 — APPROVED** [review]

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [reply src/platform_impl/android/mod.rs]: The only two relevant platforms for this function appear to be Windows and X11, where Windows has it defined here:

https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-devmodea

Seems to stem from an era where we had less than 32 bits per pixel to define a color (or palette?) and not even a single clue about the pixel _format_ (and related color space).

Let's postpone this discussion now that we can return `None` everywhere this function makes no sense.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs]: Couldn't inline it because its called in two places.
But renamed!

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs]: Done.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [inline src/platform_impl/android/mod.rs:1036]: This could panic, but is a known issue on all backends.
See https://github.com/rust-windowing/winit/issues/3258.

File: src/monitor.rs
Comment (kchibisov) [inline src/monitor.rs:66]: Let's change it to `NonZero` as well I guess, can't be zero anyway.

File: src/platform_impl/android/ffi.rs
Comment (daxpedda) [inline src/platform_impl/android/ffi.rs:1]: @MarijnS95 let me know if you think anything in here might actually panic.

**Review by kchibisov — APPROVED** [review]
I'd probably separate android impl into separate commit and rebase 2 commits in the end, but up to you.

File: src/platform_impl/android/mod.rs
Comment (kchibisov) [reply src/platform_impl/android/mod.rs:1036]: Well, if you function returns `Option` you can easily avoid panic.

Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1036]: I will make a follow-up PR where I finally tackle #3258.

Comment (kchibisov) [reply src/platform_impl/android/mod.rs:1036]: The thing is that monitor data, such as name, etc, is _static_ more or less. The position could change, indeed, but generally unless we make all of that into events and lifetime I'm not sure we can do much.

Comment (daxpedda) [reply src/platform_impl/android/mod.rs:1036]: Good point.
I'm happy to cache the name, but again, lets discuss it in a follow-up PR.

**Comment by daxpedda** [comment]
> I'd probably separate android impl into separate commit and rebase 2 commits in the end, but up to you.

I did change it up a bit to make it easier to review.
Thanks for the suggestion!

**Comment by MarijnS95** [comment]
Without going through the details, it looks like the Android implementation is quite overdone, now with JNI. Even if we discussed this before, I'd like to revise my statement.

When I proposed looking at Android functions and eventually calling them through JNI, it was meant as a hint that there appear to be matching APIs, and could eventually be used _if there is a pressing need for it_.
Using JNI also beats the user preference for a "native" activity (even though that itself is driven by some Java code, the underlying _implementations_ that you talk against are all native once they've left the `Activity` lifecycle).

---

After all, having multiple displays on Android is extremely unlikely, and even for the current display, what is the user going to use this information for? Anything involving dimensions, pixel formats, refresh rate or color spaces will _always end up going through the `Surface`_. I think Wayland and specifically its color management protocol are opting for the same thing: everything goes through the surface, and the app doesn't and shouldn't really need to know about the display layout.

---

Summary: even if we could (as you have already shown): are you sure we want to start having and managing this blob of JNI "noise" in the Android backend? For likely zero actual use-case?

**Comment by daxpedda** [comment]
> Summary: even if we could (as you have already shown): are you sure we want to start having and managing this blob of JNI "noise" in the Android backend? For likely zero actual use-case?

AFAIC I was only doing this to prevent having an `Option` on this method and learning a bit of JNI on the way.
I'm happy to remove the whole thing **but** Android should then return no `MonitorHandle` at all to prevent this issue, exactly like it was for Web before #3801.

On the other hand I've already implemented it and we will need some JNI sooner or later anyway (I assume?). I don't know if a multi-monitor setup on Android would even work with the current API (not talking about the implementation) or if it actually needs a dedicated platform-specific API (which I would assume).

Let me know how you would like to proceed.

**Comment by MarijnS95** [comment]
Besides the effort you've put in, my preference goes out to not returning this information to the user at all. In case someone requests it we can always aim to re-land your commit from this PR (let's track it somewhere just in case) or keep it around as a reference for users desiring to implement this themselves?

I am not thinking of much that needs JNI now, except maybe the PRs that are ramping up towards IME support (whether that belongs in our out `winit`, I don't know)?

Either way, I've always desired to create an ["`android-support`"](https://github.com/MarijnS95/android-support) kind of crate (under the `rust-mobile` umbrella) that abstracts away the JNI code (and could use precompiled Java/Kotlin classes for the real nasty stuff that is just too horrible to write in Rust via JNI). Details aside, we could move your JNI code there already?

**Comment by kchibisov** [comment]
I'd note that if `Android` stuff is controversial, we can split it into separate PR, since the first 3 commits should be good to go.

**Comment by daxpedda** [comment]
> Besides the effort you've put in, my preference goes out to not returning this information to the user at all. In case someone requests it we can always aim to re-land your commit from this PR (let's track it somewhere just in case) or keep it around as a reference for users desiring to implement this themselves?

I'm unsure why we would remove it in the first place if we would be willing to put it back in?
But maybe we just don't know yet *if* and how we want JNI in Winit yet?

> Either way, I've always desired to create an ["`android-support`"](https://github.com/MarijnS95/android-support) kind of crate (under the `rust-mobile` umbrella) that abstracts away the JNI code (and could use precompiled Java/Kotlin classes for the real nasty stuff that is just too horrible to write in Rust via JNI). Details aside, we could move your JNI code there already?

Yes, lets discuss it on Matrix further!

> I'd note that if `Android` stuff is controversial, we can split it into separate PR, since the first 3 commits should be good to go.

I'm gonna do the removal in a couple of hours or so.
But we should wait for somebody to test/properly review the MacOS changes.

**Comment by daxpedda** [comment]
@MarijnS95 PR is ready for your approval.

**Comment by MarijnS95** [comment]
> I'm unsure why we would remove it in the first place if we would be willing to put it back in?
> But maybe we just don't know _if_ and how we want JNI in Winit yet?

Yeah, this mostly. I don't want to commit to maintaining JNI inside `winit` now, especially given that there doesn't appear to be any usefulness in exposing Android displays and modes via `winit`. If someone comes up with a tangible use-case, we can re-evaluate, perhaps by having your code in a separate crate that's more generically usable by the Android Rust ecosystem (as proposed above, kind of like a sibling to the `ndk` crate but for Java-only interfaces), and call that from `winit` instead :+1:

File: src/platform_impl/android/mod.rs
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:976]: Just checking: was this made `Option` because of Android?

File: src/changelog/unreleased.md
Comment (MarijnS95) [inline src/changelog/unreleased.md:109]: "and `VideoModeHandle`", or is that implicit to `MonitorHandle`?

**Review by MarijnS95 — APPROVED** [review]
Android looks good!

File: src/changelog/unreleased.md
Comment (daxpedda) [reply src/changelog/unreleased.md:109]: Yes, because without a `MonitorHandle` there is no way to acquire a `VideoModeHandle`.

File: src/platform_impl/android/mod.rs
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:976]: No, Web is the only one that actually needs it so far.
Windows doesn't count because its only using `None` when the handle becomes invalid, which we have to solve differently in the future anyway.

**Pull request merged**
